projects/lesson-4/runtime/src/kitties.rs

```diff
@@ -0,0 +1,234 @@
+
+		Ok(())
+	}
+	fn do_create(owner: T::AccountId) -> Result {
+		let kitty_id = Self::next_kitty_id()?;
+
+		// Generate a random 128bit value
+		let dna = Self::random_value(&owner);
+
+		// Create and store kitty
+		
+		let kitty = Kitty(dna);
+		Self::insert_kitty(owner.clone(), kitty_id, kitty );
+		Ok(())
+	}
+
+	fn do_breed(sender: T::AccountId, kitty_id_1: T::KittyIndex, kitty_id_2: T::KittyIndex) -> Result {
+		let kitty1 = Self::kitty(kitty_id_1);
+		let kitty2 = Self::kitty(kitty_id_2);
+
+		ensure!(kitty1.is_some(), "Invalid kitty_id_1");
+		ensure!(kitty2.is_some(), "Invalid kitty_id_2");
+		ensure!(kitty_id_1 != kitty_id_2, "Needs different parent");
+
+		let kitty_id = Self::next_kitty_id()?;
+
+		let kitty1_dna = kitty1.unwrap().0;
+		let kitty2_dna = kitty2.unwrap().0;
+
+		// Generate a random 128bit value
+		let selector = Self::random_value(&sender);
+		let mut new_dna = [0u8; 16];
+
+		// Combine parents and selector to create new kitty
+		for i in 0..kitty1_dna.len() {
+			new_dna[i] = combine_dna(kitty1_dna[i], kitty2_dna[i], selector[i]);
+		}
+
+		Self::insert_kitty(sender, kitty_id, Kitty(new_dna));
+
+		Ok(())
+	}
+}
+
+
+#[cfg(test)]
+mod tests {
+	use super::*;
+
+	use runtime_io::with_externalities;
+	use primitives::{H256, Blake2Hasher};
+	use support::{impl_outer_origin, assert_ok, parameter_types};
+	use sr_primitives::{traits::{BlakeTwo256, IdentityLookup}, testing::Header};
+	use sr_primitives::weights::Weight;
+	use sr_primitives::Perbill;
+
+	impl_outer_origin! {
+		pub enum Origin for Test {}
+	}
+
+	// For testing the module, we construct most of a mock runtime. This means
+	// first constructing a configuration type (`Test`) which `impl`s each of the
+	// configuration traits of modules we want to use.
+	#[derive(Clone, Eq, PartialEq)]
+	pub struct Test;
+	parameter_types! {
+		pub const BlockHashCount: u64 = 250;
+		pub const MaximumBlockWeight: Weight = 1024;
+		pub const MaximumBlockLength: u32 = 2 * 1024;
+		pub const AvailableBlockRatio: Perbill = Perbill::from_percent(75);
+	}
+	impl system::Trait for Test {
+		type Origin = Origin;
+		type Call = ();
+		type Index = u64;
+		type BlockNumber = u64;
+		type Hash = H256;
+		type Hashing = BlakeTwo256;
+		type AccountId = u64;
+		type Lookup = IdentityLookup<Self::AccountId>;
+		type Header = Header;
+		type WeightMultiplierUpdate = ();
+		type Event = ();
+		type BlockHashCount = BlockHashCount;
+		type MaximumBlockWeight = MaximumBlockWeight;
+		type MaximumBlockLength = MaximumBlockLength;
+		type AvailableBlockRatio = AvailableBlockRatio;
+		type Version = ();
+	}
+	impl Trait for Test {
+		type KittyIndex = u32;
+	}
+	type TestModule = Module<Test>;
+
+	// This function basically just builds a genesis storage key/value store according to
+	// our desired mockup.
+	fn new_test_ext() -> runtime_io::TestExternalities<Blake2Hasher> {
+		system::GenesisConfig::default().build_storage::<Test>().unwrap().into()
+	}
+
+	#[test]
+	fn test_dna_combine() {
```

Comment:
Good to see there is test, since it's not a module function, it can be tested directly without setting up the runtime genesis config with `with_externalities`.